base/arraymath.jl

```diff
@@ -58,69 +58,83 @@ end
-    sd = size(A, d)
-    if sd == 1 || isempty(A)
-        return copy(A)
-    end
+"""
+    reverse(A; dims=:)
 
-    B = similar(A)
+Reverse `A` along dimension `dims`, which can be an integer (a
+single dimension), a tuple of integers (a tuple of dimensions)
+or `:` (reverse along all the dimensions, the default).  See
+also [`reverse!`](@ref) for in-place reversal.
 
-    nnd = 0
-    for i = 1:nd
-        nnd += size(A,i)==1 || i==d
-    end
-    if nnd==nd
-        # reverse along the only non-singleton dimension
-        for i = 1:sd
-            B[i] = A[sd+1-i]
-        end
-        return B
-    end
+# Examples
+```jldoctest
+julia> b = Int64[1 2; 3 4]
+2×2 Matrix{Int64}:
+ 1  2
+ 3  4
+
+julia> reverse(b, dims=2)
+2×2 Matrix{Int64}:
+ 2  1
+ 4  3
+
+ julia> reverse(b)
+2×2 Array{Int64,2}:
+ 4  3
+ 2  1
+```
+
+!!! compat "Julia 1.6"
+    Prior to Julia 1.6, only single-integer `dims` are supported in `reverse`.
+"""
+reverse(A::AbstractArray; dims=:) = _reverse(A, dims)
+_reverse(A, dims) = reverse!(copymutable(A); dims)
+
+"""
+    reverse!(A; dims=:)
+
+Like [`reverse`](@ref), but operates in-place in `A`.
 
-    d_in = size(A)
-    M = 1
-    for i = 1:d-1
-        M *= d_in[i]
+!!! compat "Julia 1.6"
+    Multidimensional `reverse!` requires Julia 1.6.
+"""
+reverse!(A::AbstractArray; dims=:) = _reverse!(A, dims)
+_reverse!(A::AbstractArray{<:Any,N}, ::Colon) where {N} = _reverse!(A, ntuple(identity, Val{N}()))
+_reverse!(A, dim::Integer) = _reverse!(A, (Int(dim),))
+_reverse!(A, dims::NTuple{M,Integer}) where {M} = _reverse!(A, Int.(dims))
+function _reverse!(A::AbstractArray{<:Any,N}, dims::NTuple{M,Int}) where {N,M}
+    if N < M || !allunique(dims) || !all(d -> 1 ≤ d ≤ N, dims)
+        throw(ArgumentError("invalid dimensions $dims in reverse!"))
     end
-    N = length(A)
-    stride = M * sd
-
-    if M==1
-        for j = 0:stride:(N-stride)
-            for i = 1:sd
-                ri = sd+1-i
-                B[j + ri] = A[j + i]
-            end
-        end
-    else
-        if allocatedinline(T) && M>200
-            for i = 1:sd
-                ri = sd+1-i
-                for j=0:stride:(N-stride)
-                    offs = j + 1 + (i-1)*M
-                    boffs = j + 1 + (ri-1)*M
-                    copyto!(B, boffs, A, offs, M)
-                end
-            end
-        else
-            for i = 1:sd
-                ri = sd+1-i
-                for j=0:stride:(N-stride)
-                    offs = j + 1 + (i-1)*M
-                    boffs = j + 1 + (ri-1)*M
-                    for k=0:(M-1)
-                        B[boffs + k] = A[offs + k]
-                    end
-                end
-            end
+    M == 0 && return A # nothing to reverse
+
+    idx = Vector{Int}(undef, N) # temporary array for index calculations
+
+    # compute sizes for half of reversed array
+    ntuple(k -> @inbounds(idx[k] = size(A,k)), Val{N}())
+    @inbounds idx[dims[1]] = idx[dims[1]] >> 1
```

Comment:
might as well just use `÷ 2` since the compiler is smart enough that they are equivalent   for clarity sake

Reply:
They aren't actually equivalent — `>> 1` and `÷ 2` generate different instructions in LLVM.  But they might have similar performance, and in any case this computation occurs outside the loop so it isn't performance-sensitive.

Reply:
You right these are not 100% equivalent instruction wise,  but the timings wise  they are effectively identical.

Last I checked most c compilers were smart enough to take care of such micro-optimizations making them unnecessary, but I wonder why LLVM doesn't do it in this case (if I recall correctly, on older Julia versions they compiled down to the same instructions, but I might be misremembering).